subprojects/core-api/src/main/java/org/gradle/api/artifacts/ComponentMetadataRule.java

```diff
@@ -0,0 +1,29 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.artifacts;
+
+import org.gradle.api.Action;
+import org.gradle.api.Incubating;
+
+/**
+ * A rule that modify {@link ComponentMetadataDetails component metadata}.
+ *
+ * @since 4.9
+ */
+@Incubating
+public interface ComponentMetadataRule extends Action<ComponentMetadataContext> {
```

Comment:
:+1: I like that component metadata rule is finally a thing. It's so annoying to find where you can use rules when there's only `Action` to search for!